xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-filters/xwiki-platform-notifications-filters-default/src/main/java/org/xwiki/notifications/filters/migration/R140400000XWIKI15460DataMigration.java

```diff
@@ -0,0 +1,131 @@
+import org.xwiki.notifications.NotificationException;
+import org.xwiki.notifications.filters.NotificationFilterPreference;
+import org.xwiki.notifications.filters.internal.NotificationFilterPreferenceStore;
+import org.xwiki.stability.Unstable;
+import org.xwiki.wiki.descriptor.WikiDescriptorManager;
+import org.xwiki.wiki.manager.WikiManagerException;
+
+import com.xpn.xwiki.store.migration.DataMigrationException;
+import com.xpn.xwiki.store.migration.XWikiDBVersion;
+import com.xpn.xwiki.store.migration.hibernate.AbstractHibernateDataMigration;
+
+/**
+ * Cleanup the notification filters preferences remaining on the main wiki from previously removed sub-wikis. Note: this
+ * class is named {@code R131006000XWIKI1546} in branch {@code 13.10.6+}.
+ *
+ * @version $Id$
+ * @see <a href="https://jira.xwiki.org/browse/XWIKI-15460">XWIKI-15460: Notification filter preferences are not cleaned
+ *     when a wiki is deleted</a>
+ * @since 14.4
+ * @since 13.10.6
+ */
+@Component
+@Singleton
+@Named("R140400000XWIKI1546")
+@Unstable
+public class R140400000XWIKI15460DataMigration extends AbstractHibernateDataMigration
+{
+    @Inject
+    private WikiDescriptorManager wikiDescriptorManager;
+
+    @Inject
+    private NotificationFilterPreferenceStore store;
+
+    @Inject
+    private Logger logger;
+
+    @Override
+    public XWikiDBVersion getVersion()
+    {
+        return new XWikiDBVersion(140400000);
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "Remove the notification filters preferences remaining from removed sub-wikis.";
+    }
+
+    @Override
+    public boolean shouldExecute(XWikiDBVersion startupVersion)
+    {
+        boolean shouldExecute = super.shouldExecute(startupVersion);
+        if (shouldExecute) {
+            shouldExecute = Objects.equals(this.wikiDescriptorManager.getCurrentWikiId(),
+                this.wikiDescriptorManager.getMainWikiId());
+        }
+
+        if (shouldExecute) {
+            int version = startupVersion.getVersion();
+            shouldExecute = !(version >= 131006000 && version < 140000000);
+        }
+        return shouldExecute;
+    }
+
+    @Override
+    protected void hibernateMigrate() throws DataMigrationException
+    {
+        try {
+            Collection<String> knownWikiIds = this.wikiDescriptorManager.getAllIds();
+            Set<String> unknownWikiIds = new HashSet<>();
+
+            int limit = 1000;
+            int offset = 0;
+            Set<NotificationFilterPreference> allNotificationFilterPreferences = this.store
+                .getPaginatedFilterPreferences(limit, offset);
+            while (!allNotificationFilterPreferences.isEmpty()) {
+                // We count the deleted filter preferences and adapt the next offset accordingly.
+                for (NotificationFilterPreference filterPreference : allNotificationFilterPreferences) {
+                    filterPreference.getWikiId().ifPresent(wikiId -> {
```

Comment:
This logic does not fully match the one you have in deleteFilterPreference where a preference is deleted if any of the references contains the wiki ID. But maybe it's not possible to have references with different wiki ID ?

Reply:
Afaiu wiki/page/pageOnly/user are exclusive. I'll double check though.